### PR TITLE
Fix loading native libraries from native search paths provided by the…

### LINKED ISSUE #1680: Beta8 blocker: Loading native libraries from the native search paths is not working outside Windows
Native search paths provided by the host are not searched properly outside Windows.

### PR DESCRIPTION
**Pull request by kouvel**
… host

Fixed a couple of issues outside Windows regarding the directory separator char, path separator char, and absolute path determination.

Fixes #1680

### CODE REVIEW

**Comment by kouvel** [comment]
@jkotas @sergiy-k

**Comment by davidfowl** [comment]
@kouvel Are there any tests for this code path?

File: src/inc/clr/fs/path.h [hunk 1]
```diff
@@ -55,35 +74,40 @@ namespace clr
...
-            //-----------------------------------------------------------------------------------------
             // Returns true if wzPath represents a relative path.
             static inline bool
             IsRelative(LPCWSTR wzPath)
             {
-                _ASSERTE(!clr::str::IsNullOrEmpty(wzPath)); // Must check this first.
-                return !HasLongFormatPrefix(wzPath)
-                    && !HasUncPrefix(wzPath)
-                    && (!HasDrivePrefix(wzPath) || wzPath[2] != W('\\'));
+                _ASSERTE(wzPath != nullptr);
```
Comment (moozzyk) [inline src/inc/clr/fs/path.h]: Can `wzPath` be an empty string?

**Comment by kouvel** [comment]
No it doesn't look like it, I was going to add a test separately, as that would take some more time.

File: src/inc/clr/fs/path.h [hunk 2]
```diff
@@ -55,35 +74,40 @@ namespace clr
...
+
+                // Similar to System.IO.Path.IsRelative()
+#if PLATFORM_UNIX
+                if(wzPath[0] == VolumeSeparatorChar)
+                {
+                    return false;
+                }
+#else
+                // Check for a paths like "C:\..." or "\\...". Additional notes:
```
Comment (moozzyk) [inline src/inc/clr/fs/path.h]: Will this code work for a path like "a"? You seem to be accessing [2] without checking if the string has more than 1 character...

File: src/inc/clr/fs/path.h (hunk 1 above)
Comment (kouvel) [inline src/inc/clr/fs/path.h]: Not in the paths where this function is currently called, but for the purposes of this function, an empty string is considered a relative path. The old code checking for relative path allowed for an empty string, and so does System.IO.Path.IsRelative(), so wanted to make it consistent with those.

File: src/inc/clr/fs/path.h [hunk 3]
```diff
@@ -105,7 +129,8 @@ namespace clr
                 size_t cchBuf = *pcchBuffer;
 
                 IfFailRet(StringCchCopyExW(wzBuf, cchBuf, wzPathLeft, &wzBuf, &cchBuf, STRSAFE_NULL_ON_FAILURE));
-                IfFailRet(StringCchCatExW(wzBuf, cchBuf, wzBuf[-1] == W('\\') ? W("") : W("\\"), &wzBuf, &cchBuf, STRSAFE_NULL_ON_FAILURE));
+                const WCHAR directorySeparatorWString[] = {DirectorySeparatorChar, W('\0')};
+                IfFailRet(StringCchCatExW(wzBuf, cchBuf, wzBuf[-1] == DirectorySeparatorChar ? W("") : directorySeparatorWString, &wzBuf, &cchBuf, STRSAFE_NULL_ON_FAILURE));
```
Comment (moozzyk) [inline src/inc/clr/fs/path.h]: `wzBuf[-1]` - why is it using a negative index? How does this work?

File: src/inc/clr/fs/path.h (hunk 2 above)
Comment (kouvel) [inline src/inc/clr/fs/path.h]: Before checking wzPath[2], wzPath[0] and wzPath[1] are checked here:
```c++
                if (wzPath[0] != W('\0') &&  
                    wzPath[1] == VolumeSeparatorChar && 
```

File: src/inc/clr/fs/path.h (hunk 1 above)
Comment (kouvel) [inline src/inc/clr/fs/path.h]: By old code I mean AbsolutePath() in dllimport.cpp

File: src/inc/clr/fs/path.h (hunk 2 above)
Comment (moozzyk) [inline src/inc/clr/fs/path.h]: yeah

File: src/inc/clr/fs/path.h (hunk 3 above)
Comment (kouvel) [inline src/inc/clr/fs/path.h]: StringCchCopyEx modifies wzBuf to point to the terminating null character in the destination buffer after copying the source. wzPathLeft is checked for empty above, so wzBuf will be greater than the passed-in wzBuffer. wzBuf[-1] would then get the last non-null character that was copied into the destination buffer.

**Comment by sergiy-k** [comment]
LGTM

File: src/inc/clr/fs/path.h [hunk 4]
```diff
@@ -36,6 +36,25 @@ namespace clr
         class Path
         {
         public:
+#if !PLATFORM_UNIX
+            static const CHAR DirectorySeparatorChar = '\\';
```
Comment (janvorli) [inline src/inc/clr/fs/path.h]: We already have two of these defined in palrt.h:
```C
#ifdef PLATFORM_UNIX
#define DIRECTORY_SEPARATOR_CHAR_A '/'
#define DIRECTORY_SEPARATOR_CHAR_W W('/')
#define DIRECTORY_SEPARATOR_STR_W W("/")
#define PATH_SEPARATOR_CHAR_W W(':')
#define PATH_SEPARATOR_STR_W W(":")
#else // PLATFORM_UNIX
#define DIRECTORY_SEPARATOR_CHAR_A '\\'
#define DIRECTORY_SEPARATOR_CHAR_W W('\\')
#define DIRECTORY_SEPARATOR_STR_W W("\\")
#define PATH_SEPARATOR_CHAR_W W(';')
#define PATH_SEPARATOR_STR_W W(";")
#endif // PLATFORM_UNIX
```
And palclr.h:
```C
#define DIRECTORY_SEPARATOR_CHAR_A '\\'
#define DIRECTORY_SEPARATOR_CHAR_W W('\\')
#define DIRECTORY_SEPARATOR_STR_W W("\\")

#define PATH_SEPARATOR_CHAR_W W(';')
#define PATH_SEPARATOR_STR_W W(";")
```

So I'd suggest adding VOLUME_SEPARATOR_CHAR over there and using the existing constants.

Comment (kouvel) [inline src/inc/clr/fs/path.h]: Ah ok, I'll fix this as part of a separate change (already merged this one).

**Pull request merged**